### ISSUE TITLE
sslscan Output Changes Question- Server Signature Algorithms Colours

### ISSUE THREAD

**Issue by Himself132** [rbsec/sslscan#201]
Hello,

Long time user of sslscan here, love the tool.  In doing a comparison in my most recent security assessment of previous findings while running the tool I noticed some great new features.  Looking at what sslscan does in contrast to ssllabs which is only available for public-facing internet structure I commonly use the output from both sources to come up with a write-up on available and problematic SSL/TLS cipher suites and other associated cryptographic implementations and controls.

I see that you are flagging some of the Server Signature Algorithms the server is showing as available as problematic.  I do not see a corresponding output from ssllabs (not that it matters).  Based on my understanding of this field it is a transitionary 1.3 capability that originated with TLS 1.2 but whose purpose may have changed some and potentially introduced some additional security concerns.  It sounds like if a client only has one cipher suite available and advertises that it prompts a server to respond with information about what options there are, which may even not be compatible with the certificate being validated?  What is the attack vector here that is triggering the coloring?  Is it just because the server does have a weak Server Signature Algorithm and if it were to have other certificates that it could be a problem?

In short - Can you explain a scenario (flow-wise) where the negotiation could go awry and a Server Signature Algorithm that the server has available could be used by a client in some way controllable or influenced by a malicious attacker?

**Comment by jtesta** [comment]
@Himself132 : I wrote the server signature algorithm enumeration code, so I can give some insight to how it works.

This test will perform a handshake with the TLS endpoint with each known server signature type (as defined in [RFC5246](https://tools.ietf.org/html/rfc5246#section-7.4.1.4.1)) listed as acceptable by the client.  If the server responds with a ServerHello response, this means the server is--_so far_--OK with this algorithm.  This doesn't necessarily mean that the server must reply with this signature type later, only that it wasn't rejected immediately.

The rationale behind this test is that a good endpoint should have all bad signature types disabled outright, so as to prevent old & outdated client software from silently creating faulty TLS tunnels.  Similarly to reducing the cipher list to only good ciphersuites, if the server is configured to only allow good parameters, then it would be impossible for bad connections to be created on accident.

The threat model involving a malicious client intentionally creating faulty tunnels is not relevant in this scenario.

I hope this helps!

**Comment by rbsec** [comment]
As @jtesta says, this is more of a defence-in-depth control than a directly exploitable issue (in as much as anything TLS related is directly exploitable). There are a few different scenarios that you might be when you scan a server:

* The server only allows good/modern signature algorithms - this means that any attempt to use old and weaker algorithms in a connection will fail.
* The server allows old/weak signature algorithms - this means that it _might_ be possible to negotiation a connection using those, but both sides would have to agree for this to happen.
* The server allows all signature algorithms/ignores this field - same as above.

If the output from sslscan indicated that the server supports `rsa_pkcs1_sha1` (for example), this means that it _might_ be possible to use this algorithm during the connection (although it is unlikely that both sides will negotiate to use this unless they don't support anything better) - which could leave the connection more vulnerable to various cryptographic attacks compared to algorithms. However, it could also be the case that the server will refuse to use these algorithms at another stage - so allowing it here doesn't necessarily mean that it would ever be used.

It's worth nothing that the majority of servers do not allow you to configure the supported signature algorithms - it's possible if you're directly using the OpenSSL library, but as far as I know it's not possible to change these in Apache/Nginx/IIS/etc.

**Comment by Himself132** [comment]
@rbsec @jtesta - Thank you for the information.  Maybe I am missing something but based on what i can see in the research i've done.  The server signature algorithms aren't available for use by an agreement between the client and server unless the certificate was signed with the same algorithm.  Is that not correct?  Based on what you have stated above it sounds like there would be downgrade attack capabilities if not done securely.  For instance below, wouldn't the signature algorithm in the certificate as shown below (sha256WithRSAEncryption) have to match rsa_pkcs1_sha256 inthe list and all else would fail as the certificate is not utilizing another server signature algorithm?

`  Server Signature Algorithm(s):
TLSv1.2  rsa_pkcs1_sha1
TLSv1.2  dsa_sha1
TLSv1.2  ecdsa_sha1
TLSv1.2  rsa_pkcs1_sha224
TLSv1.2  dsa_sha224
TLSv1.2  ecdsa_sha224
TLSv1.2  rsa_pkcs1_sha256
TLSv1.2  dsa_sha256
TLSv1.2  ecdsa_secp256r1_sha256
TLSv1.2  rsa_pkcs1_sha384
TLSv1.2  dsa_sha384
TLSv1.2  ecdsa_secp384r1_sha384
TLSv1.2  rsa_pkcs1_sha512
TLSv1.2  dsa_sha512
TLSv1.2  ecdsa_secp521r1_sha512

  SSL Certificate:
Signature Algorithm: sha256WithRSAEncryption
RSA Key Strength:    2048
`

So basically, if a downgrade attack were available would this even work if the agreement was made between both parties after such an attack IF it were possible?

**Comment by jtesta** [comment]
> The server signature 
> algorithms aren't available for use by an agreement between the client 
> and server unless the certificate was signed with the same algorithm. Is 
> that not correct?

The CA's signature on the server's certificate (done once) is separate 
from the signing that the server does to set up each new TLS connection.

For example, the CA key could be RSA 4096, and the server's key could be 
DSA 1024.  The CA hashes the server's DSA key with SHA-256 and signs it 
with the CA's RSA key, generating the sha256WithRSAEncryption signature 
type; this is embedded within the certificate, which never changes. 
When a new client comes in, the server then uses its DSA key to sign 
various parts of the connection setup.  It can directly use DSA with 
SHA-1, SHA-256, or SHA-512 (or any hash algorithm approved in an RFC 
somewhere).

Furthermore, its possible for the server to generate a temporary ECDSA 
or ED25519 keypair, sign the public portion of it with its DSA key, and 
send it off to the client.  From that point on, the server could sign 
subsequent messages with ECDSA/ED25519 and not its long-term DSA key. 
(I've got some brain rust on this last part, so don't quote me on it.  I 
could use a refresher...)

Hence, the server can use all manners of key types that aren't directly 
hard-coded in its certificate.

-- 
Joseph S. Testa II
Founder & Principal Security Consultant
Positron Security

**Comment by Himself132** [comment]
Thank you! @rbsec can you add anything?  Specific to the downgrade attack vectors?  Say if someone were to man-in-the-middle and exploit a TLS implementation would this be a viable attack?

**Comment by jtesta** [comment]
> Specific to the downgrade attack vectors?

There are no known downgrade attacks between TLS 1.1 through 1.3.  And 
I'm 90% sure updated TLS 1.0 implementations aren't affected either.


> Say if someone were to 
> man-in-the-middle and exploit a TLS implementation would this be a 
> viable attack?

There are no known man-in-the-middle attacks on the protocols themselves.

However, if you allow end users to make decisions, they'll often make 
the wrong one (this is the key behind my unrelated ssh-mitm project). 
Last I checked, Chrome strongly dissuades the user from accepting 
invalid certificates.

-- 
Joseph S. Testa II
Founder & Principal Security Consultant
Positron Security
